d3log/src/tcp_network.rs

```diff
@@ -0,0 +1,114 @@
+use tokio::{net::TcpListener,
+            net::TcpStream,
+            io::AsyncReadExt,
+            io::AsyncWriteExt};
+
+use core::future::Future;
+use std::net::{SocketAddr, IpAddr, Ipv4Addr};
+use crate::{json_framer::JsonFramer, Node, Batch, transact::TransactionManager};
+use differential_datalog::ddval::DDValConvert;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::pin::Pin;
+use async_std::sync::Mutex;
+use crate::child::output_json;
+use crate::batch::singleton;
+
+
+async fn process_socket(_n:&TcpNetwork, mut s: TcpStream) -> Result<(), std::io::Error> {
+    let mut jf = JsonFramer::new();
+    let mut buffer = [0; 64];
+    let k = s.read(&mut buffer).await?;
+    
+    for i in jf.append(&buffer[0..k])? {
+        println!("{}", i);
+    }
+    Ok(())
+}
+
+
+pub fn address_to_nid(peer:SocketAddr) -> Node{
+    let i = match peer.ip(){
+        IpAddr::V4(x) => x.octets(),
+        IpAddr::V6(_) => return 0
+    };
+    let p = peer.port().to_be_bytes();    
+
+    // prefer big endian for generic reasons, but we are travelling through
+    // a constricted u64 pipe
+    u128::from_le_bytes([i[0], i[1], i[2], i[3], p[0], p[1], 0, 0, 0, 0, 0,
+                         0,0,0,0,0])
+}
+
+pub fn nid_to_socketaddr(n:Node) -> SocketAddr {
+    let b = n.to_be_bytes();
+    SocketAddr::new(IpAddr::V4(Ipv4Addr::new(b[0], b[1], b[2], b[3])), u16::from_be_bytes([b[4], b[5]]))
+}
+
+pub struct TcpNetwork {
+    peers : Arc::<Mutex::<HashMap<Node, Arc::<Mutex::<TcpStream>>>>>
```

Comment:
The double-synchronization here is fairly suspect to me

Reply:
most of this spurious Arc-Mutex stuff Is me flailing around w/ the borrow checker. 

in this case its not completely clear though that one would have to hold the peer lock over any usage of a connection. I also don't know if reading and writing from different threads on the same socket needs to be locked (it kinda shouldn't..neither should calls to write(), but there may be additional internal state)